### PR TITLE
fix: honor updater opt-out labels during image scans

### LINKED ISSUE #3302: 🐞 Bug: 403/401 errors with containers that have build scripts
### Bug Description

Constant errors in the Arcane app and web interface.

I have a Caddy Proxy container which is a local build and it’s created / configured outside of Arcane.

No matter what I try, I cannot get Arcane to stop trying and failing to search for updates for it and showing errors in recent artifices as a result.

### Steps To Reproduce

Configure a container with a build script outside of Arcane (local source)

Try to disable auto update check

Receive errors constantly in the Arcane interface.

### Expected Behavior

Configure a local build container and have a flag available to actually have it not check for updates.

### Actual Behavior

Errors.

### Screenshots

Image

### Arcane Version

v2.4.0

### Installation Method

Docker Compose (Recommended)

### Environment Type

Not sure

### Database Type

SQLite (Default)

### Operating System

Linux

### Docker Version

29.6.0, build fb59821

### Browser & Version

_No response_

### Relevant Logs or Error Messages

```shell
local/caddy:latest: daemon digest lookup failed; registry fallback failed: distribution inspect failed for docker.io/local/caddy:latest: Error response from daemon: <html><body><h1>403 Forbidden</h1>
Request forbidden by administrative rules.
</body></html>
registry manifest inspect failed for docker.io/local/caddy:latest: authenticated manifest request failed with status: 401
```

### Docker Compose Configuration

```yaml
services:
  caddy:
    # This points to the Dockerfile in the same directory
    build: .
    image: local/caddy:latest
    container_name: caddy
    restart: unless-stopped
    ports:
      - "10.0.0.1:80:80"
      - "10.0.0.1:443:443"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - caddy_data:/data
    environment:
      - CADDY_INGRESS_NETWORKS=caddy
      - CLOUDFLARE_API_TOKEN=
    labels:
      caddy.email: "@"
      caddy.acme_dns: "cloudflare {env.CLOUDFLARE_API_TOKEN}"
      com.getarcaneapp.arcane.updater: "false"
    networks:
      caddy:

networks:
  caddy:
    external: true

volumes:
  caddy_data:

———

Build:

FROM caddy:builder AS builder

# This step downloads and compiles the custom plugins into Caddy
RUN xcaddy build \
    --with github.com/lucaslorentz/caddy-docker-proxy/v2 \
    --with github.com/caddy-dns/cloudflare

FROM caddy:latest
COPY --from=builder /usr/bin/caddy /usr/bin/caddy
CMD ["caddy", "docker-proxy"]
```

### Environment Configuration (.env file)

```shell

```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Kstateag**
## Checklist

* [x] This PR is **not** opened from my fork’s `main` branch
* [x] All new user-facing strings are translated via Paraglide (`m.*()`)

## What This PR Implements

Fixes Arcane image update scans ignoring the `com.getarcaneapp.arcane.updater=false` container label.

`CheckAllImages` previously enumerated all tagged Docker images without considering whether those images were used only by containers that had opted out of updater participation. This caused locally built images, such as `local/caddy:latest`, to be checked against a registry and produce avoidable 401/403 errors.

Fixes #3302

## Changes Made

* Added container-aware filtering to `CheckAllImages`.
* Excludes images used exclusively by containers with updater participation disabled.
* Preserves images shared by both opted-out and eligible containers.
* Excludes all repository tag aliases belonging to an exclusively opted-out Docker image.
* Falls back to normalized image-reference matching when Docker image IDs are unavailable or do not align.
* Keeps unused tagged images eligible for image update checks.
* Applies the image limit after opt-out filtering.
* Gives image and container discovery independent Docker API timeout contexts.
* Leaves explicit single-image and selected-image update checks unchanged.
* Replaced the superseded image-reference deduplication helper with coverage for the new filtering behavior.

## Testing Done

* Added regression coverage for:

  * images used exclusively by opted-out containers
  * enabled containers
  * images shared by opted-out and enabled containers
  * unused tagged images
  * limits applied after filtering
  * repository tag aliases of opted-out images
  * shared image aliases
  * normalized-reference fallback when Docker image IDs differ
* Ran focused image discovery tests:

  * `GOEXPERIMENT=jsonv2 go test ./internal/services -run 'TestImageUpdateService_GetAllImageRefs' -count=1`
* Ran the full service test package:

  * `GOEXPERIMENT=jsonv2 go test ./internal/services -count=1`
* Ran GolangCI-Lint:

  * `golangci-lint run -c ../.github/.golangci.yml ./internal/services/...`
  * Result: `0 issues`
* Ran `git diff --check`.
* Repository pre-commit formatting checks passed.

## AI Tool Used (if applicable)

ChatGPT was used to help trace the image update scan path, identify edge cases, design regression tests, and review the final implementation. All changes were manually validated with focused tests, the full service test package, repository formatting checks, and GolangCI-Lint.

## Additional Context

The updater execution path already honored the opt-out label. The defect was isolated to the image discovery path used by scheduled and manual “check all images” operations.

Explicit checks for a single image or a selected set of images still run when directly requested.

## Summary Update

This PR makes check-all image discovery container-aware by filtering images used exclusively by updater-disabled containers, while preserving shared and unused images.

Both critical issues from the original review are now resolved:

- ID-based and normalized-reference usage are merged before eligibility is decided, so an opted-out container cannot hide an image still used by an eligible container.
- Container-discovery failures now log a warning and fall back to discovered image references instead of aborting check-all scans.

The implementation uses a merge-based usage model, degrades gracefully when container listing fails, and is covered by eight focused regression tests.

**Revised Confidence Score: 4/5**

The logic is sound, the fallback is safe and properly logged, and the key scenarios are covered comprehensively. The remaining confidence gap is limited to external `refs.NormalizeImageUpdateRef` and `refs.IsImageIDLikeReference` behavior that could not be fully verified from the diff alone.

**Disclaimer** Greptiles Reviews use AI, make sure to check over its work.

To better help train Greptile on our codebase, if the comment is useful and valid **Like** the comment, if its not helpful or invalid **Dislike**

To have Greptile Re-Review the changes, mention `greptileai`.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR makes check-all image discovery honor container updater opt-out labels.
- Filters images used exclusively by opted-out containers while retaining shared and unused images.
- Merges image-ID and normalized-reference usage before determining eligibility.
- Falls back to the previously discovered image references when container discovery fails.
- Applies scan limits after opt-out filtering and adds focused regression coverage.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

The PR appears safe to merge.

Both previously reported failures are fixed: ID-based and reference-based eligibility are merged before filtering, and container-list errors retain discovered image references instead of aborting the scan; no blocking failure remains.
</details>

<sub>Reviews (2): Last reviewed commit: ["Fix Image Scan Updater Opt-Out"](https://github.com/getarcaneapp/arcane/commit/086bb556bf48b49155183ae71b0dd2184520005b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=47465982)</sub>

**Context used:**

- Knowledge Base — [Container & Image Lifecycle](https://app.greptile.com/ofkm/-/custom-context/knowledge-base/getarcaneapp/arcane/-/docs/container-image-lifecycle.md)

### CODE REVIEW

File: backend/internal/services/image_update_service.go [hunk 1]
```diff
@@ -616,35 +618,97 @@ func (s *ImageUpdateService) getAllImageRefsInternal(ctx context.Context, limit
...
 	if err != nil {
 		return nil, errors.WrapIf(err, "failed to list Docker images")
 	}
 
-	return dedupeImageRefsFromSummariesInternal(imageList.Items, limit), nil
+	containerCtx, cancelContainers := s.dockerAPIContextInternal(ctx)
+	containerList, err := dockerClient.ContainerList(containerCtx, client.ContainerListOptions{All: true})
+	cancelContainers()
+	if err != nil {
+		return nil, errors.WrapIf(err, "failed to list Docker containers")
```
[the remark this answers is not in the document]
Comment (Kstateag) [reply backend/internal/services/image_update_service.go]: @greptileai Good Catch, both flagged issues have been resolved, please reevaluate and then amend your summary and your confidence score.

File: backend/internal/services/image_update_service.go [hunk 2]
```diff
@@ -616,35 +618,97 @@ func (s *ImageUpdateService) getAllImageRefsInternal(ctx context.Context, limit
...
-				return imageRefs[:limit]
+			if _, exists := seen[imageRef]; exists {
+				continue
+			}
+
+			usage := imageUsage
+			hasUsage := hasImageUsage
+			if !hasUsage {
+				normalizedRef := refs.NormalizeImageUpdateRef(imageRef)
+				usage, hasUsage = usageByRef[normalizedRef]
```
[the remark this answers is not in the document]
Comment (Kstateag) [reply backend/internal/services/image_update_service.go]: @greptileai Good Catch, issue resolved.

**Review by kmendell — APPROVED** [review]
LGTM, Thanks
